etcdnaming/etcdnaming.go

```diff
@@ -0,0 +1,143 @@
+package etcdnaming
+
+import (
+	"fmt"
+	"sync"
+
+	etcdcl "github.com/coreos/etcd/client"
+	"golang.org/x/net/context"
+)
+
+type kv struct {
+	key, value string
+}
+
+// recvBuffer is an unbounded channel of *kv.
+type recvBuffer struct {
+	c        chan *kv
+	mu       sync.Mutex
+	stopping bool
+	backlog  []*kv
+}
+
+func newRecvBuffer() *recvBuffer {
+	b := &recvBuffer{
+		c: make(chan *kv, 1),
+	}
+	return b
+}
+
+func (b *recvBuffer) put(r *kv) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.stopping {
+		return
+	}
+	b.backlog = append(b.backlog, r)
+	select {
+	case b.c <- b.backlog[0]:
+		b.backlog = b.backlog[1:]
+	default:
+	}
+}
+
+func (b *recvBuffer) load() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.stopping || len(b.backlog) == 0 {
+		return
+	}
+	select {
+	case b.c <- b.backlog[0]:
+		b.backlog = b.backlog[1:]
+	default:
+	}
+}
+
+func (b *recvBuffer) get() <-chan *kv {
+	return b.c
+}
+
+func (b *recvBuffer) stop() {
+	b.mu.Lock()
+	b.stopping = true
+	close(b.c)
+	b.mu.Unlock()
+}
+
+type etcdNR struct {
+	kAPI   etcdcl.KeysAPI
+	recv   *recvBuffer
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// NewETCDNR creates an etcd NameResolver
+func NewETCDNR(cfg etcdcl.Config) *etcdNR {
```

Comment:
This needs to be restructured like:
naming/naming.go (type Resolver interface)
naming/etcd/etcd.go (etcd impl)

You can follow what credentials + oauth does as an example.